src/ResultWriter/ClusteringWriter.cpp

```diff
@@ -0,0 +1,73 @@
+#include "ClusteringWriter.h"
+
+#include <filesystem>
+#include <fstream>
+
+#include <Parallel/MPI.h>
+namespace seissol::writer {
+
+ClusteringWriter::ClusteringWriter(const std::string& outputPrefix) : outputPrefix(outputPrefix) {
+
+}
+
+void ClusteringWriter::addCluster(unsigned profilingId,
+                                  unsigned localClusterId,
+                                  LayerType layerType,
+                                  unsigned size,
+                                  unsigned dynRupSize) {
+  clusteringInformation.profilingIds.push_back(profilingId);
+  clusteringInformation.localClusterIds.push_back(localClusterId);
+  clusteringInformation.layerTypes.push_back(layerType);
+  clusteringInformation.sizes.push_back(size);
+  clusteringInformation.dynamicRuptureSizes.push_back(dynRupSize);
+}
+
+void ClusteringWriter::write() const {
+  using namespace std::filesystem;
+  const auto& mpi = MPI::mpi;
+
+  const auto localRanks = mpi.collect(mpi.sharedMemMpiRank());
+  const auto profilingIds = mpi.collectContainer(clusteringInformation.profilingIds);
+  const auto localClusterIds = mpi.collectContainer(clusteringInformation.localClusterIds);
+  const auto layerTypes = mpi.collectContainer(clusteringInformation.layerTypes);
+  const auto sizes = mpi.collectContainer(clusteringInformation.sizes);
+  const auto dynamicRuptureSizes = mpi.collectContainer(clusteringInformation.dynamicRuptureSizes);
+
+  if (mpi.rank() == 0) {
+
+    auto filepath = path(outputPrefix);
+    filepath += path("-clustering.csv");
+
+    auto fileStream = std::fstream(filepath, std::ios::out);
+
+    fileStream << "profilingId,localId,layerType,size,dynamicRuptureSize,rank,localRank\n";
+
+    for (int rank = 0; rank < mpi.size(); ++rank) {
+      const auto localRank = localRanks[rank];
+      const auto& curProfilingIds = profilingIds[rank];
+      const auto& curLocalClusterIds = localClusterIds[rank];
+      const auto& curLayerTypes = layerTypes[rank];
+      const auto& curSizes = sizes[rank];
+      const auto& curDynamicRuptureSizes = dynamicRuptureSizes[rank];
+
+      for (unsigned i = 0; i < curProfilingIds.size(); ++i) {
+        const auto layerType = static_cast<LayerType>(curLayerTypes[i]);
+        const auto layerTypeStr = layerType == Interior ? "Interior" : "Copy";
```

Comment:
Are we sure `Ghost` is strictly not possible?

Reply:
Yes. Ghost clusters are created seperately and are basically an abstraction for communication. (They don't have any data!)